tests/queries/0_stateless/02947_read_in_reverse_order_final.sql

```diff
@@ -0,0 +1,100 @@
+SET final=1;
+SET optimize_read_in_order=1;
+
+DROP TABLE IF EXISTS t_02947_reverse_order_final;
+
+CREATE TABLE t_02947_reverse_order_final (x int, y int, z int)
+ENGINE = ReplacingMergeTree ORDER BY (x, y);
+
+-- Level-0 part with duplicate rows
+INSERT INTO t_02947_reverse_order_final SETTINGS optimize_on_insert = 0 VALUES (0, 0, 0), (0, 0, 1);
+
+SELECT * FROM t_02947_reverse_order_final FINAL ORDER BY x DESC, y DESC LIMIT 1;
+
+INSERT INTO t_02947_reverse_order_final SETTINGS optimize_on_insert = 0 VALUES (0, 0, 0), (0, 0, 1), (0, 0, 2);
+
+SELECT * FROM t_02947_reverse_order_final FINAL ORDER BY x DESC, y DESC LIMIT 1;
+
+TRUNCATE t_02947_reverse_order_final;
+
+INSERT INTO t_02947_reverse_order_final SELECT number, number, number FROM numbers(10000);
+INSERT INTO t_02947_reverse_order_final SELECT number, number * 2, number FROM numbers(10000);
+INSERT INTO t_02947_reverse_order_final SELECT number, number * 2, number * 3 FROM numbers(10000);
+
+SELECT '---';
+SELECT if(explain like '%ReadType: InReverseOrder%', 'Ok', 'Error: ' || explain) FROM (
+    EXPLAIN PLAN actions = 1 SELECT * FROM t_02947_reverse_order_final FINAL ORDER BY x DESC, y, z LIMIT 1
+    SETTINGS max_threads = 1, max_final_threads = 1
+) WHERE explain like '%ReadType%';
+
+SELECT '---';
+SELECT count() FROM t_02947_reverse_order_final FINAL;
+SELECT * FROM t_02947_reverse_order_final FINAL ORDER BY x, y, z LIMIT 1;
+SELECT * FROM t_02947_reverse_order_final FINAL ORDER BY x DESC, y, z LIMIT 1;
+SELECT * FROM t_02947_reverse_order_final FINAL ORDER BY x, y DESC, z LIMIT 1;
+SELECT * FROM t_02947_reverse_order_final FINAL ORDER BY x DESC, y DESC, z LIMIT 1;
+SELECT * FROM t_02947_reverse_order_final FINAL ORDER BY x DESC, y DESC, z DESC LIMIT 1;
+
+DROP TABLE t_02947_reverse_order_final;
+
+CREATE TABLE t_02947_reverse_order_final (x int, y int, ver int)
+ENGINE = ReplacingMergeTree(ver) ORDER BY x;
+
+-- Level-0 part with duplicate rows
+INSERT INTO t_02947_reverse_order_final SETTINGS optimize_on_insert = 0 VALUES (0, 0, 0), (0, 1, 0), (1, 0, 0), (1, 1, 1);
```

Comment:
let's add a similar test, but each row will be inserted separately (to form 4 distinct parts).
I am a little concerned about these lines: https://github.com/ClickHouse/ClickHouse/pull/58361/files#diff-c2317ca02dca707ab63917057efaf36ca941e500948fc64857e1b7b0f29bff51R127-R128
but probably everything is ok, since we continue to process parts themselves from left to right